### DISCUSSION TITLE
What's the SMARTS format for generic or markush reaxys-style queries?

### DISCUSSION THREAD

**Discussion by ljmartin** [rdkit/rdkit#6342] [Q&A]
Hi all, 
Referring to the [RDKit book here](https://www.rdkit.org/docs/RDKit_Book.html#generic-markush-queries-in-substructure-matching). 

I'd like to match one (or two even!) heterocyclic rings as part of a SMARTS query. I'm not sure this is possible with regular SMARTS (without specifying the sizes of the rings in advance), so the generic/markush style queries look great. The SMARTS looks nonstandard, though:

```
q = Chem.MolFromSmarts('OC* |$;;ARY$|')
```
if this is not Daylight SMARTS, is there a format defined somewhere? If not: 
- what are the pipes doing? 
- what are the dollar signs doing?
- if I remove the space it fails to parse. What is the space doing? 
- likewise the `*` (there isn't a wildcard atom) 
- Can I add atoms on the other end of the ARY, or is it purely a sidechain? 

Thanks a lot - looking forward to using it!

**Comment by ljmartin** [comment]
By the way, here's the real use-case I'm after. I'm looking to match multiple heterocycles, which may be joined, inside a SMARTS pattern that might specify other things, such as alkyl groups. So something like this would be recognised as having an alkyl chain, then two heterocyclic groups (fused), then another alykl chain - the generic query would be useful because, for example, the ring sizes could be swapped and it would still match:

 
<img width="601" alt="Screenshot 2023-05-06 at 7 31 34 am" src="https://user-images.githubusercontent.com/8262765/236572214-5234e981-0822-4957-96c0-fe6ef48ba581.png">

**Reply by syedzayyan** [reply]
https://www.daylight.com/dayhtml_tutorials/languages/smarts/

I am following this today because I am in a similar boat. Maybe this helps?

**Reply by bp-kelley** [reply]
The RDKit does support some ChemAxon style queries, but you currently need to use the Generalized Substructures module:

```
from rdkit import Chem
from rdkit.Chem import rdGeneralizedSubstruct

mol = Chem.MolFromSmiles("OCC1=NC=C2N=C(C)NC2=C1")
pattern_mol = Chem.MolFromSmarts('OC* |$;;ARY$|')
xqm = rdGeneralizedSubstruct.CreateExtendedQueryMol(pattern_mol)
matches = rdGeneralizedSubstruct.MolGetSubstructMatches(mol, xqm)
mol.__sssAtoms = matches[0] # to show in Jupyter notebook...
mol
```

![image](https://github.com/rdkit/rdkit/assets/9888416/562708af-1a4b-4a57-9fdc-49de7bc8f160)

However, this doesn't do exactly what you want with wild cards.  What we generally use are mol bundles which contain a collection of smarts patterns for detecting repeated patterns, however you can easily roll your own in python.  Stealing regular expression notation, what you are looking for is "CO([a]+)C". where [a]+ is one or more  aryl atoms (this is the smarts little a, big A is aliphatic).  So we can just make the bundle as follows:

```
b = Chem.MolBundle()
bundle = []
for repeats in range(1,9):  # allow 1 to 8 aryl atoms
   smarts = "OC" + "[a]" * repeats + "C"
   b.AddMol(Chem.MolFromSmarts(smarts))
    
mol.GetSubstructMatch(b)
mol
```

![image](https://github.com/rdkit/rdkit/assets/9888416/7c2ba22c-72ea-4efc-a901-8f140f3e255a)

Hope this helps.

**Reply by bp-kelley** [reply]
(Note these examples are for Jupyter notebooks which automatically shows the highlighting for substruct matches)
